### ISSUE TITLE
[Network] Operate a retrier for multiple tasks on retry.

### ISSUE THREAD

**Issue by huitseeker** [MystenLabs/narwhal#567]
Our `BoundedExecutor` in #559 gained the ability to queue tasks on an infinite retry (with exponential backoff) so that the backoff timers a run off of the semaphore that limits its concurrency.

However, those retries are driven by one task each (see the `TODO` on `spawn_with_retry`):
https://github.com/MystenLabs/narwhal/blob/7e49fa42e32437c4039f4ae40957e241dc0c2ec1/network/src/bounded_executor.rs#L131-L168

This results in one new task being unconditionally spawned per enqueued message with retries. What should rather happen is that `BoundedExecutor` should carry one overall task that would manage retries for all such queued retrying messages (i.e. a scheduler).

**Comment by erwanor** [comment]
:wave: @huitseeker , I would like to take this on if this isn't on your hot path. Feel free to assign this to me if you want.

The design I have in mind to address this issue, happy to get feedback on it:

We create a retry task that takes in a `mpsc::Receiver<Future>` and maintains a priority queue of jobs indexed by their exponential backoff value. The retry task polls the queue for the job with highest priority, pops it from the queue if it is runnable, and attempts to drive it to completion. If it fails, the job is pushed into the queue with an updated backoff value. This mirrors the `Poll::Pending` / `Poll::Ready` approach that tokio takes. 

The drawback of this approach are that retries are blocking and stack up on one other. I think this can create something similar to a [cache stampede](https://en.wikipedia.org/wiki/Cache_stampede) scenario when many tasks are queued for retry with similar backoffs, and the first take to be driven to completion takes a long time to execute. 

An alternative non-blocking approach: instead of blocking on execution of the current task, we spawn the job in its own task, pass it some context like the `RetryConfig`, as well as a `mpsc` channel to the retry task executor. The spawned task collects the result and either winds down or register a new job to be ran by the retrier. This is probably more robust, but introduces more complexity to the bounded executor. Personally, I'm leaning towards the first approach.

What are your thoughts on this? I can start implementing once you greenlight this.

**Comment by huitseeker** [comment]
Thanks @erwanor, I appreciate your thinking on this!

This `BoundedExecutor` with retries (#463 #559) is used for sending messages. The messages we are sending using this are driving important parts of our protocol and if they are queued sequentially, one blocking sending attempt (the execution of a future) blocking too many others risks blocking the whole protocol. So blocking is not an option, I'm glad you brought that up. 

Our setup is more to have the "retry manager" not execute the futures itself, but launch a limited number of concurrent attempts on the semaphore of the `BoundedExecutor`:
https://github.com/MystenLabs/narwhal/blob/84db6c88f9a179ab3e145ecf81a86117cb522348/network/src/bounded_executor.rs#L128-L135
Today that's [by default 200](https://github.com/MystenLabs/narwhal/blob/b3728f80dd3e467c27d1b5426d95e2beeaf60178/network/src/lib.rs#L60) concurrent messages sends, and the hope is that these 200 messages will be enough to cycle through all the messages -though we will for sure revise that value as we introduce more message recipients in the protocol.

The problem, however, is that we launch one "retry manager" task, _off_ the semaphore, per "message-that-must-be-retried" (call to `spawn_with_retries`): 
https://github.com/MystenLabs/narwhal/blob/84db6c88f9a179ab3e145ecf81a86117cb522348/network/src/bounded_executor.rs#L174

If there are 200 such messages with distinct recipients, each of them needing to be retried on failure, we would launch 400 tasks (half of them on the semaphore).

So all we're trying to do here is to mutualize the retry management into one single task (200 on the semaphore, 1 managing the "queueing" of further attempts).

I'm somewhat OK with complexity in the `BoundedExecutor` in order to perform this. 

I believe your approach in managing a queue is great -with the caveat that when a future pops off the queue, it's spawned on the executor rather than executed in a blocking fashion. Have you considered using the [tokio `DelayQueue`](https://docs.rs/tokio-util/latest/tokio_util/time/struct.DelayQueue.html) for this? 

Moreover, for managing the computation of the exponential backoff priorities with a lighter dependency than the backoff crate, I note we could use the simple [exponential-backoff](https://github.com/yoshuawuyts/exponential-backoff).

**Comment by erwanor** [comment]
@huitseeker great write-up, thanks for your feedback on this. This is helpful!

Based on your feedback, and perusing the docs, here's some pseudocode:

First, we define a tuple `Job` that wraps a task to be executed, and retried on failure, as well as a backoff context that is specific to that task. The backoff gets "ticked" on every failure. 

```rust
struct Job<F, T, E>(F, Backoff); // F: FnMut () -> Future<Output = Result<T, E>> [...]; 

```
Conceptually, we have a `RetryManager` owning a `DelayQueue<Job>`, and an inbound mpsc channel `Receiver<Job>` to register new jobs for execution. In practice, avoiding nesting `RetryManager` into the `BoundedExecutor` might turn out cleaner. I have to write some more code to figure that out.

```rust
struct RetryManager {
    /// A queue of failed tasks indexed by their backoff cooldown
    job_queue: DelayQueue<Job>,
    /// A sink to submit new jobs to the scheduler
    retry_job_tx: Arc<mpsc::Sender<Job>>,
    /// A stream of new jobs to register for retry execution
    retry_job_rx: mpsc::Receiver<Job>,
}
```

Depending on what backoff crate we are working with,  we would either change or massage the `BoundedExecutor` API so that we have:

```rust
impl BoundedExecutor {
     // [...]
  pub(crate) fn spawn_with_retries<F, Fut, T, E>(
        &self,
        backoff: Backoff,
        mut f: F,
    ) -> JoinHandle<Result<T, E>>
    where
        F: FnMut() -> Fut + Send + 'static,
        Fut: Future<Output = Result<T, E>> + Send + 'static,
        T: Send + 'static,
        E: Send + 'static,
    {
            let task = f();
            let retry_tx = self.retry_jobs_tx.clone();
            let retry_task = task.map(|result| async move {
                match result {
                     Ok(_) => (),
                     Err(_) => {
                         let backoff = backoff.tick();
                         retry_tx.send(Job(f, backoff)).await;
                     }
                result
           });
           self.spawn(retry_task);
    }
```

Next, we have the actual `retry_manager` task whose lifetime is tied to the `BoundedExecutor`. It pulls jobs from either a `DelayQueue` (ready to be executed), or from a mpsc channel (to be scheduled).

```rust
// retry_manager logic:
loop {
     select! {
          // Pull a job that's ready to be retried
          job_ready = job_queue.next() => {
                  let (f, backoff) = job_ready.split();
                  let task = f();
                  // chain future with retry logic; doesn't compile, consider this pseudocode
                  let retry_job_tx = self.tx.clone(); // Arc<mpsc::Sender<Job>>
                  let retry_task = task.map(|result| move async {
                        match result {
                            Ok(_) => (),
                            Err(_) => {
                                let backoff = backoff.tick();
                                let job = Job(f, backoff);
                                retry_job_tx.send(job).await;
                           };
                           result
                  }).flatten();  // We can do a simple closure too

                  // Spawns the task on the BoundedExecutor's semaphore
                  // **Blocking** until a permit is acquired: instead, should we queue the task 
                  // for execution and use a second worker to drive those to completion?
                  //  Instead of blocking on `spawn` we would block on a buffered channel 
                  //  send. The advantage being that the `retry_manager` can still make
                  //  progress on scheduling (i.e. inserting in the `delay_queue`) and 
                  //  awaiting the `send` creates backpressure when the queue is full.
                  self.spawn(retry_task).await
          }, 
          // A job is marked for retry execution.
          job = new_jobs.recv() => {
                    let (f, backoff) = job.split();
                    let cooldown_time = backoff.time();
                    delay_queue.insert(Job(f, backoff), cooldown_time);
                 },
}
```

I have pointed out a couple possible flaws with this draft, in particular, the `retry_manager` is doing a blocking `spawn` on the semaphore. I suggest a way to mitigate that in my comment. I didn't implement it so that we can focus on the overall mechanic of the `retry_manager`. I'm also wondering if the retry tasks and "normal" tasks should share a semaphore, or each have their own.

This might be a pathological case but the combination of a high transient failure rate and a too optimistic backoff policies (exponential, but ultimately specified by the client) could create starvation in the bounded executor and trickle down into liveness issues. I'm just thinking out loud, and this might be out-of-scope for this specific issue. 

thanks again for your feedback, looking forward to your reply!

**Comment by huitseeker** [comment]
Thanks @erwanor for the sketch of your approach. This makes sense to me. 

One question: the main problem I see with the blocking send is not so much the blocking send (we can live with inaccuracies in the timing of the retries, and will tune the number of permits in the semaphore), but that it prevents checking other branches of the `tokio::select`. As a general rule, we do not like to hide memory consumption in unbounded `mpsc` channels, which means that unscheduled jobs could pile up until the channel to the retry scheduler is full.

here are some thoughts on mitigation:
- to only spawn a job when the `new_jobs` receiver is empty? See https://github.com/MystenLabs/narwhal/pull/724 for an approach that simultaneously polls a future for an upcoming work item along with another that checks an async gating function is true as well,
- to make `spawn_with_retries` output a `Result` in order to pass the backpressure on to the caller.

But I'm 100% satisfied by neither.

**Comment by erwanor** [comment]
Ah good point, thanks for catching that. Then I think we can drop the idea of a separate runner and block on a buffer [mpsc::Sender::send](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.send). And instead, use a blocking spawn, but gate the branch on acquiring a permit to the semaphore first, and pass it to the executor later.  If we still want to go with a separate execution task and a `send.await`, we can gate that branch on a `reserve()` like you do in `try_fut_and_permit!`.  

>to make spawn_with_retries output a Result in order to pass the backpressure on to the caller.

If I understand correctly, we would return a `Result<JoinHandle, BoundedExecutionError>`. So technically, a `try_spawn_with_retries` correct? I think that's quite complementary to how we handle bp in the scheduler. It would be neat to give clients the ability to poll for readiness without submitting tasks, or committing resources in the `BoundedExecutor`, but that's only a nice-to-have. 

I'm opening a draft PR so that we have something concrete to iterate on! I'll update it as we narrow down on a design. Thanks again for the feedback.

**Comment by erwanor** [comment]
@huitseeker 👋 , I completed a big chunk of the scheduler logic, and re-organized the `BoundedExecutor` to accomodate the `RetryManager`'s `'static` task, without changing the contract for the clients. The code is still a WIP in #801, but at a high-level, here's the implementation:

```
///   BoundedExecutor
/// ┌───────────────────────────────────────────────────────────────────────────────┬────────────┐
/// │                                                                               │            │
/// │                                      tx_new_jobs                              │ DelayQueue │
/// │                   ┌────────────────────────────────────────────────────┐      │(exec_queue)│
/// │                   │                                                    │      │            │
/// │                   │                                                    │      │            │
/// │                   │                                                    │  ┌──►│  insert    │
/// │    ┌──────────────┴────────────┐    ┌─────────────────────┐            │  │   │            │
/// │    │                           │    │                     │            │  │   │            │
/// │    │                           │    │                     │            │  │   │            │
/// │    │ .InternalBoundedExecutor  │    │     RetryManager    │            ▼  │   │            │
/// │    │                           │    │                     │               │   │            │
/// │    │ ┌──────────┬────────────┐ │    │                     │     ┌─────────┴─┐ │            │
/// │    │ │          │            │ │    │                     │     │ scheduler │ │            │
/// │    │ │  Handle  │  Semaphore │ │    │                     │ run │           │ │            │
/// │    │ │ (runtime)│            │ │    │                     ├────►│           │ │            │
/// │    │ │          │            │ │    │                     │     │           │ │            │
/// │    │ │          │            │ │    │                     │     │           │ │            │
/// │    │ │          │            │ │    │                     │     └──────┬────┘ │            │
/// │    │ │          │            │ │    │                     │            │  ▲   │            │
/// │    │ │          │            │ │    │                     │            │  │   │            │
/// │    │ │          │            │ │    │                     │            │  │   │            │
/// │    │ └──────────┴────────────┘ │    │                     │            │  │   │            │
/// │    │                           │    │                     │            │  │   │            │
/// │    └───────────────────────────┘    └─────────────────────┘            │  └───┤ job_ready  │
/// │                  ▲                                                     │      │            │
/// │                  └─────────────────────────────────────────────────────┘      │            │
/// │                                    spawn                                      │            │
/// │                                                                               │            │
/// └───────────────────────────────────────────────────────────────────────────────┴────────────
```

In the current design, the `BoundedExecutor` is split into:
- an inner `Arc<InternalBoundedExecutor>` which encapsulates the logic, runtime, and semaphore.
- a `CancelOnDropHandler` to a long-running scheduler task consuming a `RetryManager` (labeled `scheduler`)

The flow of execution looks like this:
- A task is spawned with `spawn_with_retries`
- On failure, a `Job` is sent to the schedule. `Job` wraps a `FaillibleFuture` factory, and a `Backoff` context. (`tx_new_jobs`)
- The scheduler queues the job for execution (`insert` in queue).
- After polling the execution queue, and contingent on acquiring resources, the job is ran on the `BoundedExecutor` semaphore. (`job_ready` -> `spawn_with_permit`)

There remains the open question of surfacing results. I'm wondering if you had thoughts about this. I'm thinking`spawn_with_retries` could return a future that polls a `oneshot::channel` rather than a`JoinHandle` over a running task. A possible implementation then is to return a future that polls a callback `oneshot` channel, and maybe have some mechanic to acquire / release permits early on `Poll::Pending`, and have the waker await a semaphore permit. Thinking out loud here. I think there's a tradeoff here. - please let me know what you think!

**Comment by huitseeker** [comment]
I think the overall design makes sense to me. 
- Surfacing results to the outside of the `BoundedExecutor`, through an output value of some sort is not a priority in this case: we assume the tasks either are silent or embark their own side-effecting signaling (e.g. a oneshot channel) to produce their output. Failures will log. That's all,
- OTOH, those very same futures should very much signal (inside the `Executor`) whether it's worth retrying them through their output. [What the backoff crate does](https://docs.rs/backoff/latest/backoff/enum.Error.html), which seems sensible to me is they ask the futures to be `TryFuture`s, and to wrap their `Error` case in a `backoff::Error` with two variants, `Transient` (which indicates a retry should happen) and `Permanent` (which indicates a retry will not succeed). Requiring something analogous to this this through trait bounds is perfectly reasonable to me.

**Comment by erwanor** [comment]
@huitseeker makes sense, that's implemented now!

For the last mile between polling the retry task and returning a cancel-on-drop handler, I have extended the `BoundedExecutor` API with [`spawn_with_retries_and_adapter`](https://github.com/MystenLabs/narwhal/pull/801/files#diff-947e9d4ccfc31a9565e0ef85e0010709a9382581ff8259ac26887ba20a2383d7R251).

This method lets clients specify an adapter converting `Result<(), RetryError>` into a `Result<T, E>` _before_ spawning is done, such that spawning returns a `JoinHandle<Result<T, E>>`.  My reasoning is detailed in the commit message [3fdab80fe](https://github.com/MystenLabs/narwhal/pull/801/commits/3fdab80fe279f3b409631f795df8ed8e59e92760), but to summarise: this gets us cleaner type signatures, and fits nicely with the existing network traits (`-> CancelOnDropHandler<MessageResult>`) without dynamic dispatch or excessively allocating memory in the manager.

When you get the chance, please let me know what you think!

**Comment by erwanor** [comment]
I am attacking `select!` branch gating as well as writing tests.  My initial approach of doing a naive `join!` on `acquire_permit` and `DelayQueue::poll_expired` won't work. It turns out polling an empty queue returns a `Poll::Ready(None)`. This creates repeated branch misfires when the queue is empty and a permit is available.

I'm considering wrapping the `DelayQueue` to turn change the polling behavior:  [that's how it looks like](https://github.com/erwanor/patientqueue/blob/main/src/lib.rs#L22). This isn't ideal, and I have other leads as well. I'd be curious to hear your thoughts on this, open to all options.
